src/helpers/gateway/send.rs

```diff
@@ -55,47 +97,64 @@ impl<M: Message> SendingEnd<M> {
     pub async fn send(&self, record_id: RecordId, msg: M) -> Result<(), Error> {
-        if let TotalRecords::Specified(count) = self.total_records {
-            if usize::from(record_id) >= count.get() {
-                return Err(Error::TooManyRecords {
-                    record_id,
-                    channel_id: self.channel_id.clone(),
-                    total_records: self.total_records,
-                });
-            }
-        }
-
+        let r = self.inner.send(record_id, msg).await;
         metrics::increment_counter!(RECORDS_SENT,
             STEP => self.channel_id.step.as_ref().to_string(),
-            ROLE => self.my_role.as_static_str()
+            ROLE => self.sender_role.as_static_str()
         );
 
-        self.ordering_tx
-            .send(record_id.into(), Wrapper::wrap(&msg))
-            .await
-            .map_err(|e| Error::send_error(self.channel_id.clone(), e))
+        r
     }
 }
 
 impl GatewaySenders {
     /// Returns or creates a new communication channel. In case if channel is newly created,
     /// returns the receiving end of it as well. It must be send over to the receiver in order for
     /// messages to get through.
-    pub(crate) fn get_or_create(
+    pub(crate) fn get_or_create<M: Message>(
         &self,
         channel_id: &ChannelId,
         capacity: NonZeroUsize,
-    ) -> (
-        OrderingMpscSender<Wrapper>,
-        Option<OrderingMpscReceiver<Wrapper>>,
-    ) {
+        total_records: TotalRecords, // TODO track children for indeterminate senders
+    ) -> (Arc<GatewaySender>, Option<GatewaySendStream>) {
+        assert!(!total_records.is_unspecified());
         let senders = &self.inner;
         if let Some(sender) = senders.get(channel_id) {
-            (sender.clone(), None)
+            (Arc::clone(&sender), None)
         } else {
-            let (tx, rx) = ordering_mpsc::<Wrapper, _>(format!("{channel_id:?}"), capacity);
-            senders.insert(channel_id.clone(), tx.clone());
-            (tx, Some(rx))
+            const SPARE: Option<NonZeroUsize> = NonZeroUsize::new(64);
+            // a little trick - if number of records is indeterminate, set the capacity to 1.
+            // Any send will wake the stream reader then, effectively disabling buffering.
+            // This mode is clearly inefficient, so avoid using this mode.
+            let write_size = if total_records.is_indeterminate() {
+                capacity
+            } else {
+                NonZeroUsize::new(1).unwrap()
+            };
+
+            let sender = Arc::new(GatewaySender::new(
+                channel_id.clone(),
+                OrderingSender::new(write_size, SPARE.unwrap()),
+                total_records,
+            ));
+            if senders
+                .insert(channel_id.clone(), Arc::clone(&sender))
+                .is_some()
+            {
+                panic!("TODO - make sender creation contention less dangerous");
```

Comment:
I want to flag this as a potentially very bad bug.  I've put a `panic!()` here, but it might need some more careful thought instead.